wpiutil/src/test/native/cpp/rotated_span_test.cpp

```diff
@@ -0,0 +1,204 @@
+    EXPECT_EQ(sp[5], arr_values[5]);
+  }
+  {
+    wpi::rotated_span<const int> sp{cvec_values.begin(), cvec_values.end()};
+    EXPECT_EQ(sp[5], vec_values[5]);
+  }
+  {
+    wpi::rotated_span<const int> sp{cvec_values.data(), cvec_values.size()};
+    EXPECT_EQ(sp[5], vec_values[5]);
+  }
+}
+
+TEST(CircularSpanTest, ConstructNonConst) {
+  {
+    wpi::rotated_span<int, 10> sp{sarr_values};
+    EXPECT_EQ(sp[5], sarr_values[5]);
+  }
+  {
+    wpi::rotated_span<int> sp{sarr_values};
+    EXPECT_EQ(sp[5], sarr_values[5]);
+  }
+  {
+    wpi::rotated_span<int, 10> sp{arr_values};
+    EXPECT_EQ(sp[5], arr_values[5]);
+  }
+  {
+    wpi::rotated_span<int> sp{arr_values};
+    EXPECT_EQ(sp[5], arr_values[5]);
+  }
+  {
+    wpi::rotated_span<int> sp{vec_values.begin(), vec_values.end()};
+    EXPECT_EQ(sp[5], vec_values[5]);
+  }
+  {
+    wpi::rotated_span<int> sp{vec_values.data(), vec_values.size()};
+    EXPECT_EQ(sp[5], vec_values[5]);
+  }
+}
+
+TEST(CircularSpanTest, ConstructRotated) {
+  {
+    constexpr wpi::rotated_span<const int, 10> sp{cesarr_values, 1};
+    static_assert(sp[5] == cesarr_values[6]);
+  }
+  {
+    constexpr wpi::rotated_span<const int, 10> sp{cesarr_values, 9};
+    static_assert(sp[5] == cesarr_values[4]);
+  }
+  {
+    constexpr wpi::rotated_span<const int, 10> sp{cesarr_values, 10};
+    static_assert(sp[5] == cesarr_values[5]);
+  }
+  {
+    constexpr wpi::rotated_span<const int, 10> sp{cesarr_values, 11};
+    static_assert(sp[5] == cesarr_values[6]);
+  }
+
+  {
+    constexpr wpi::rotated_span<const int, 10> sp{cearr_values, -1};
+    static_assert(sp[5] == cearr_values[4]);
+  }
+  {
+    constexpr wpi::rotated_span<const int, 10> sp{cearr_values, -9};
+    static_assert(sp[5] == cearr_values[6]);
+  }
+  {
+    constexpr wpi::rotated_span<const int, 10> sp{cearr_values, -10};
+    static_assert(sp[5] == cearr_values[5]);
+  }
+  {
+    constexpr wpi::rotated_span<const int, 10> sp{cearr_values, -11};
+    static_assert(sp[5] == cearr_values[4]);
+  }
+}
+
+TEST(CircularSpanTest, Rotate) {
+  constexpr wpi::rotated_span<const int, 10> sp{cesarr_values, 1};
+  static_assert(sp[5] == cesarr_values[6]);
+  static_assert(sp.rotate(2)[5] == cesarr_values[8]);
+  static_assert(sp.rotate(9)[5] == cesarr_values[5]);
+  static_assert(sp.rotate(10)[5] == cesarr_values[6]);
+  static_assert(sp.rotate(11)[5] == cesarr_values[7]);
+  static_assert(sp.rotate(-1)[5] == cesarr_values[5]);
+  static_assert(sp.rotate(-2)[5] == cesarr_values[4]);
+  static_assert(sp.rotate(-9)[5] == cesarr_values[7]);
+  static_assert(sp.rotate(-10)[5] == cesarr_values[6]);
+  static_assert(sp.rotate(-11)[5] == cesarr_values[5]);
+}
+
+void unsized_func(wpi::rotated_span<int>) {}
+void const_unsized_func(wpi::rotated_span<const int>) {}
+void sized_func(wpi::rotated_span<int, 10>) {}
+void const_sized_func(wpi::rotated_span<const int, 10>) {}
+
+TEST(CircularSpanTest, Implicit) {
+  // unsized_func(csarr_values);  // error
```

Comment:
A bit clunky, but if we want to make sure these are invalid, we could define a concept that checks if a call with a parameter is valid (using a requires expression) and then use static asserts:
```cpp
template <auto Param, auto Func>
concept ValidParamFor = requires {
  Func(Param);
};

static_assert(!ValidParamFor<csarr_values, unsized_func>);
```
Note that we do need to define the concept, since "If a substitution failure would occur in a requires-expression for every possible template argument, the program is ill-formed, no diagnostic required." ([requires expression](https://en.cppreference.com/w/cpp/language/requires))